crates/sui-graphql-rpc/src/extensions/query_limits_checker.rs

```diff
@@ -38,6 +51,182 @@ struct QueryLimitsCheckerExt {
+                                        self.max_tx_payload_size
+                                    ),
+                                    f.pos
+                                ));
+                        }
+                        let var_name = f
+                            .node
+                            .arguments
+                            .iter()
+                            .map(|x| x.1.node.to_string())
+                            .collect::<Vec<_>>()
+                            .concat();
+                        // the key is the txBytes+signatures value itself or the variables' names
+                        track_vars.entry(var_name).or_insert(tx_bytes_len);
+                    }
+                }
+            }
+        }
+
+        let cfg = ctx
+            .data::<ServiceConfig>()
+            .expect("No service config provided in schema data");
+
+        // remove the executeTransactionBlock and the dryRunTransactionBlock variables from the
+        // total query size, and check if the rest is bigger than the allowed limit
+        if (payload_size - track_vars.values().sum::<usize>() as u64)
+            > cfg.limits.max_query_payload_size.into()
+        {
+            self.log_metric(ctx, "Query payload is too large: {}", payload_size);
+            return Err(graphql_error(
+                code::BAD_USER_INPUT,
+                format!(
+                    "Query payload is too large. The maximum allowed is {} bytes",
+                    cfg.limits.max_query_payload_size
+                ),
+            ));
+        }
+
+        Ok(())
+    }
+
+    /// Check the body content-length against the allowed payload. The content-length will include
+    /// variables data as well.
+    fn check_overall_payload_size(&self, ctx: &ExtensionContext<'_>) -> ServerResult<()> {
+        let payload_size: &PayloadSize = ctx.data_unchecked();
+        let payload_size = payload_size.0;
+        let max_payload_size = (self.max_query_payload_size + self.max_tx_payload_size).into();
+        if payload_size > max_payload_size {
+            let metrics: &Metrics = ctx.data_unchecked();
+            let query_id: &Uuid = ctx.data_unchecked();
+            let session_id: &SocketAddr = ctx.data_unchecked();
+            metrics
+                .request_metrics
+                .query_payload_too_large_size
+                .observe(payload_size as f64);
+            info!(
+                query_id = %query_id,
+                session_id = %session_id,
+                error_code = code::BAD_USER_INPUT,
+                "Query payload is too large: {}",
+                payload_size
+            );
+            return Err(graphql_error(
+                code::BAD_USER_INPUT,
+                format!("Query payload is too large. Max allowed is {max_payload_size}"),
+            ));
+        }
+        Ok(())
+    }
+
+    fn log_metric(&self, ctx: &ExtensionContext<'_>, message: &str, payload_size: u64) {
```

Comment:
There's no use of the `self` parameter here AFAICT -- and given the second, simpler check may not be necessary, I think that tips it so that the entire check can be turned into a single function, rather than introducing the type for it.